tpm.go

```diff
@@ -0,0 +1,937 @@
+	// TPM for which the endorsement certificate was issued, so a correct response means we're communicating with that TPM.
+	symmetric := tpm2.SymDef{
+		Algorithm: tpm2.SymAlgorithmAES,
+		KeyBits:   tpm2.SymKeyBitsU{Data: uint16(128)},
+		Mode:      tpm2.SymModeU{Data: tpm2.SymModeCFB}}
+	session, err := t.StartAuthSession(ek, nil, tpm2.SessionTypeHMAC, &symmetric, defaultSessionHashAlgorithm, nil)
+	if err != nil {
+		return xerrors.Errorf("cannot create HMAC session: %w", err)
+	}
+	succeeded := false
+	defer func() {
+		if succeeded {
+			return
+		}
+		t.FlushContext(session)
+	}()
+
+	if len(t.verifiedEkCertChain) > 0 {
+		_, err = t.GetRandom(20, session.WithAttrs(tpm2.AttrContinueSession|tpm2.AttrAudit))
+		if err != nil {
+			if isAuthFailError(err) {
+				return verificationError{errors.New("endorsement key proof of ownership check failed")}
+			}
+			return xerrors.Errorf("cannot execute command to complete EK proof of ownership check: %w", err)
+		}
+	}
+
+	succeeded = true
+
+	if ekIsPersistent {
+		t.ek = ek
+	}
+	t.hmacSession = session
+	return nil
+}
+
+// readEkCertFromTPM reads the manufacturer injected certificate for the default RSA2048 EK from the standard index, and
+// returns it as a DER encoded byte slice.
+func readEkCertFromTPM(tpm *tpm2.TPMContext) ([]byte, error) {
+	ekCertIndex, err := tpm.CreateResourceContextFromTPM(ekCertHandle)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot create context: %w", err)
+	}
+
+	ekCertPub, _, err := tpm.NVReadPublic(ekCertIndex)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot read public area of index: %w", err)
+	}
+
+	cert, err := tpm.NVRead(ekCertIndex, ekCertIndex, ekCertPub.Size, 0, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot read index: %w", err)
+	}
+
+	return cert, nil
+}
+
+// openDefaultTcti connects to the default TPM character device. This can be overridden for tests to connect to a simulator device.
+var openDefaultTcti = func() (io.ReadWriteCloser, error) {
+	return tpm2.OpenTPMDevice(tpmPath)
+}
+
+// connectToDefaultTPM opens a connection to the default TPM device.
+func connectToDefaultTPM() (*tpm2.TPMContext, error) {
+	tcti, err := openDefaultTcti()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot open TPM device: %w", err)
+	}
+
+	tpm, _ := tpm2.NewTPMContext(tcti)
+	return tpm, nil
+}
+
+// checkChainForEkCertUsage checks whether the specified certificate chain is valid for a TPM endorsement key certificate.
+// For any certificate in the chain that defines extended key usage, that key usage must either be anyExtendedKeyUsage (RFC5280),
+// or it must contain tcg-kp-EKCertificate.
+func checkChainForEkCertUsage(chain []*x509.Certificate) bool {
+	if len(chain) == 0 {
+		return false
+	}
+
+NextCert:
+	for i := len(chain) - 1; i >= 0; i-- {
+		cert := chain[i]
+		if len(cert.ExtKeyUsage) == 0 && len(cert.UnknownExtKeyUsage) == 0 {
+			continue
+		}
+
+		for _, usage := range cert.ExtKeyUsage {
```

Comment:
(nitpick) with a small helper like `isExtKeyUsageAny()` this would probably be slightly more readable and avoids the continue to label.

Reply:
Yeah, that makes it a bit neater - I've added isExtKeyUsageAny and isExtKeyUsageEkCertificate as well.